---
title: "Summary"
output: html_document
params:
  folder: /lustre/scratch126/cellgen/cellgeni/pasham/raw/2110.SNARE-seq/v04.fq/pipeline.out/F2_msBrain_Nu
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(visutils)
library(ArchR)
library(Rsamtools)
library(Seurat)
library(Matrix)
```

# Working directory
```{r}
print(params$folder)
```

# Load Data
```{r}
qc = readRDS(paste0(params$folder,'/qc.rds'))
ataca =readRDS(paste0(params$folder,'/whole.archr.prj.rds'))
```

# Load Metrics
```{r}
cdna.metrics_ = cdna.metrics = qc$metrics$cdna.metrics
atac.metrics_ = atac.metrics = qc$metrics$atac.metrics
```

```{r}
# prepare for printing
cdna.metrics_[['Valid barcodes no mismatches']]   = paste0(cdna.metrics[['Valid barcodes no mismatches']] ,' (',round(cdna.metrics[['Valid barcodes no mismatches' ]]/cdna.metrics[['Total reads']]*100,1),'%)')
cdna.metrics_[['Valid barcodes <=1 mismatches']]  = paste0(cdna.metrics[['Valid barcodes <=1 mismatches']],' (',round(cdna.metrics[['Valid barcodes <=1 mismatches']]/cdna.metrics[['Total reads']]*100,1),'%)')
cdna.metrics_[['Uniquely mapped reads']]          = paste0(round(cdna.metrics[['Uniquely mapped reads']]*100,1),'%')
cdna.metrics_[['GeneFull unique reads']]          = paste0(round(cdna.metrics[['GeneFull unique reads']]*100,1),'%')
cdna.metrics_[['Gene unique reads']]              = paste0(round(cdna.metrics[['Gene unique reads']]*100,1),'%')
cdna.metrics_[['Fraction MT UMI']]                = paste0(round(cdna.metrics[['Fraction MT UMI']]*100,2),'%')


atac.metrics_[['mean TSS enrichment score']]                  = round(atac.metrics[['mean TSS enrichment score']],1)
atac.metrics_[['Valid barcodes no mismatches']]               = paste0(atac.metrics[['Valid barcodes no mismatches']] ,' (',round(atac.metrics[['Valid barcodes no mismatches' ]]/atac.metrics[['Read pairs']]*100,1),'%)')
atac.metrics_[['Valid barcodes <=1 mismatches']]              = paste0(atac.metrics[['Valid barcodes <=1 mismatches']],' (',round(atac.metrics[['Valid barcodes <=1 mismatches']]/atac.metrics[['Read pairs']]*100,1),'%)')
atac.metrics_[['Fragments in cells']]                         = paste0(atac.metrics[['Fragments in cells']],' (',round(atac.metrics[['Fragments in cells']]/atac.metrics[['Total fragments used']]*100,1),'%)')
atac.metrics_[['Fraction MT reads']]                          = paste0(round(atac.metrics[['Fraction MT reads']]*100,2),'%')  
atac.metrics_[['Fraction non-chr reads']]                     = paste0(round(atac.metrics[['Fraction non-chr reads']]*100,3),'%')
atac.metrics_[['Fraction of genome in peaks']]                = paste0(round(atac.metrics[['Fraction of genome in peaks']]*100,1),'%')
atac.metrics_[['Fraction of fragments overlapping TSS']]      = paste0(round(atac.metrics[['Fraction of fragments overlapping TSS']]*100,1),'%')
atac.metrics_[['Fraction of fragments overlapping peaks']]    = paste0(round(atac.metrics[['Fraction of fragments overlapping peaks']]*100,1),'%')
atac.metrics_[['Fraction duplicates']]                        = paste0(round(atac.metrics[['Fraction duplicates']]*100,1),'%')
```


# General info
```{r}
r = data.frame(names(qc$metrics$info),unname(unlist(qc$metrics$info)))
colnames(r) = NULL#c(paste(rep('_',max(nchar(r[,1]),na.rm=T)+5),collapse = ''),paste(rep('_',max(nchar(r[,2]),na.rm=T)+5),collapse = ''))
knitr::kable(r)
```

# ATAC metrics
```{r}
r = data.frame(names(atac.metrics_),unname(unlist(atac.metrics_)))
colnames(r) = NULL
knitr::kable(r)
```

# cDNA metrics
```{r}
r = data.frame(names(cdna.metrics_),unname(unlist(cdna.metrics_)))
colnames(r) = NULL
knitr::kable(r)
```


# Cell calling
```{r, fig.height = 5, fig.width = 5}
cells = ataca@cellColData$cdna_cell
plot(ataca@cellColData$nFrags+1,ataca@cellColData$cdna_total_umi+1,log='xy',pch=16,bty='n',xlab='ATAC fragments',ylab='cDNA UMIs',col=cells+1,cex=0.5)
legend('topleft',bty='n',col=1:2,legend=paste0(c('Non-Cells','Cells'),' (',c(sum(!cells),sum(cells)),')'),pch=16)
```

# ATAC QC
```{r,fig.height = 5, fig.width = 10}
par(mfrow=c(1,2),bty='n')
plot(qc$fragsize$fragmentSize,qc$fragsize$fragmentPercent,xlab='ATAC-seq Fragment Size (bp)',ylab='Percentage of Fragments',type='l',lwd=2)
plot(qc$tssenrich$x,qc$tssenrich$smoothValue,xlab='Distance From Center (bp)',ylab='Normalized Intensity Profile',type='l',lwd=2)
#plot(atacf@cellColData$nFrags,atacf@cellColData$FRIP,pch=16,log='x')
```

# Cell Clustering
```{r,fig.height = 10, fig.width = 14}
par(mfrow=c(2,2),mar=c(0,0,1,10),bty='n')
cex = min(1/nrow(qc$cdna.umap)*3000,1)
plotVisium(qc$cdna.umap,qc$cdna.umap$seurat_clusters,main='cDNA, clusters by cDNA',t='xy',pch=16,show.cluster.sizes = TRUE,label.clusters = TRUE,randomize.points = TRUE,cex=cex)
plotVisium(qc$cdna.umap,qc$atac.umap$Clusters,main='cDNA, clusters by ATAC',t='xy',pch=16,show.cluster.sizes = TRUE,label.clusters = TRUE,randomize.points = TRUE,cex=cex)

plotVisium(qc$atac.umap,qc$cdna.umap$seurat_clusters,main='ATAC, clusters by cDNA',t='xy',pch=16,show.cluster.sizes = TRUE,label.clusters = TRUE,randomize.points = TRUE,cex=cex)
plotVisium(qc$atac.umap,qc$atac.umap$Clusters,main='ATAC, clusters by ATAC',t='xy',pch=16,show.cluster.sizes = TRUE,label.clusters = TRUE,randomize.points = TRUE,cex=cex)
```


# UMI/fragment coverage
```{r,fig.height = 5, fig.width = 14}
par(mfrow=c(1,2),mar=c(0,0,1,10),bty='n')
plotVisium(qc$cdna.umap,qc$cdna.umap$nFeature_RNA,zfun = log1p,main='cDNA',t='xy',pch=16,legend.args = list(title='total UMIs'),randomize.points = TRUE,cex=cex)
plotVisium(qc$atac.umap,qc$atac.umap$nFrags,zfun = log1p,main='ATAC',t='xy',pch=16,legend.args = list(title='total frags'),randomize.points = TRUE,cex=cex)
```


```{r,fig.height = 5, fig.width = 14}
par(mfrow=c(1,2),mar=c(0,0,1,10),bty='n')
plotVisium(qc$atac.umap,qc$atac.umap$FRIP,main='ATAC (fraction reads in peaks)',t='xy',pch=16,randomize.points = TRUE,cex=cex)
plotVisium(qc$atac.umap,qc$atac.umap$TSSEnrichment,main='ATAC (TSSEnrichment)',t='xy',pch=16,randomize.points = TRUE,cex=cex)
```



```{r,fig.height = 5, fig.width = 8}
t = table(qc$atac.umap$Clusters,qc$cdna.umap$seurat_clusters)
t = t[,order(apply(t,2,which.max))]
main = paste0('Chisq p=',format(chisq.test(t)$p.value,scienciefic=TRUE,digits=2))
imageWithText(log1p(t),t,xlab='clusters by ATAC',ylab='clusters by cDNA',col=num2col(1:100,c('white','gray','orange','red')),bty='n',main=main)
```